### PR TITLE
GitHub Actions Workflow to replace Travis CI

### LINKED ISSUE #209: Using GitHub Actions -- replace Travis CI integration?
I've created the YAML script to drive GitHub Actions (GA) testing, in commit c7630ae + 3b10324 (the latter because I didn't realize before that GA cancels everything in the matrix if any one of them fails.)

As you can [see here](https://github.com/aio-libs/aiosmtpd/actions), the tests passes for the current HEAD of `master`.

(You can manually trigger it to run by clicking on "aiosmtpd CI" on the left sidebar, then click "Run workflow".)

I don't know how to integrate them to the PR system, though. And/or how to replace Travis CI automatic testing of PR and pushes-to-PR.

### PR DESCRIPTION
**Pull request by pepoluan**
## What do these changes do?

Configures GitHub Actions, and stop using Travis CI.

Also banishes the very annoying & irritating "intermittent coverage failure" error that has been plaguing the PRs since, like, forever.

## Are there changes in behavior for the user?

None whatsoever. This is repo back-end only.

## Related issue number

Closes #209 

## Checklist

(I've removed irrelevant checklist items)

- [x] I think the code is well written

### CODE REVIEW

File: .github/workflows/test.yml
Comment (pepoluan) [inline .github/workflows/test.yml:29]: GitHub Actions documentation guarantees that `CI == "true"`

See https://docs.github.com/en/free-pro-team@latest/actions/reference/environment-variables#default-environment-variables

**Comment by pepoluan** [comment]
Hmmm ... no trigger? Let's try closing & reopening

**Comment by pepoluan** [comment]
Stupidity was mine. Misplaced a key config item.

**Comment by pepoluan** [comment]
Sorry for the noise ... I can't possibly test GitHub Actions on my laptop.

File: .github/workflows/test.yml
Comment (pepoluan) [inline .github/workflows/test.yml:52]: Not sure if `always()` is needed or not... just in case it's needed, we add it here so codecov.io will receive reports of 100% coverage _and_ <100% ones.

File: tox.ini
Comment (pepoluan) [inline tox.ini:14]: Moved to `cov` part to make the "codecov.io reporting" step of GitHub Actions work without running `diffcov`.

Now that I think of it ... XML report is really actually part of cov's report generation.

`diffcov` merely post-process cov's report, is all.

So this is a very proper move.

File: .github/workflows/test.yml
Comment (pepoluan) [inline .github/workflows/test.yml:13]: A bit confused at `continue-on-error` description in GA's docs ... seems like the whole workflow can still succeed if any matrix fail?

So I used a more explicit "allow one matrix to fail" setting instead: `fail-fast` (see below)

File: .github/workflows/test.yml
Comment (pepoluan) [inline .github/workflows/test.yml:19]: When one matrix combination fails, do NOT cancel other matrix combinations. Let them all run to completion.

**Review by pepoluan — COMMENTED** [review]
Some explanations so far.

**Comment by pepoluan** [comment]
Well I'll be ... it all passes!

**Review by pepoluan — COMMENTED** [review]
It's done. Travis is gone.

**Comment by waynew** [comment]
Have we confirmed that tests not passing will fail/block the PR? If not, just doing a quick change in here to intentionally fail would probably be good.

**Comment by pepoluan** [comment]
> Have we confirmed that tests not passing will fail/block the PR? If not, just doing a quick change in here to intentionally fail would probably be good.

Good idea. I'll try intentionally pushing a faliling test later today.

GA did behave like Travis CI, though, in the regards that falling tests did result in red ❌s appearing in the list of checks. But since the PR was in Draft, there was no Merge button.

**Comment by pepoluan** [comment]
The reason I force pushed is because I'm yet to have my first cup of coffee and accidentally amended a commit... 😅

**Comment by pepoluan** [comment]
image

👆 @waynew when GA check fails, the Merge button becomes gray (instead of green). But it's still clickable, probably because I have write access to the repo?

**Comment by pepoluan** [comment]
As a comparison, from my p.o.v., **Travis failure resulted in the same:**

image

"Merge button" becomes gray (no longer green), but **still clickable**.

(This one's tested on PR #202 )

**Comment by waynew** [comment]
Yep, that'll do it :+1:

I'll endeavor to give this a thorough look through later today!

File: .github/workflows/test.yml
Comment (asvetlov) [reply .github/workflows/test.yml:52]: https://github.com/codecov/codecov-action is easier to configure IMHO. There is no need to check these matrix values anymore.

**Comment by pepoluan** [comment]
> 
> 
> Yep, that'll do it 👍
> 
> I'll endeavor to give this a thorough look through later today!

Awesome!

Do you still need the "all fail" state, or may I revert the intentional test fail?

(I'll try adopting fc65a84 , the solution to the "intermittent coverage failure" from #202 to unittest, so that everything will ["should"] pass... GitHub Actions seems to be more prone to this issue than Travis.)

**Comment by waynew** [comment]
You can go ahead and revert it :+1: I just wanted to make sure we at least tested that failures would fail :joy:

**Comment by pepoluan** [comment]
frikkin dammit I forgot to do `tox -e qa` before committing & pushing 😑

**Comment by pepoluan** [comment]
Amended (previous) last commit, force pushed, and now it's green across the board, yay! 😄

File: aiosmtpd/smtp.py
Comment (pepoluan) [inline aiosmtpd/smtp.py:246]: No longer needed thanks to `test_starttls:TestTLSEnding:test_eof_received`, which works on Windows.

File: aiosmtpd/smtp.py
Comment (pepoluan) [inline aiosmtpd/smtp.py:254]: No longer relevant because we've removed the `# pragma: nomswin` marker there.

File: aiosmtpd/tests/test_starttls.py
Comment (pepoluan) [inline aiosmtpd/tests/test_starttls.py:88]: "Secret sauce" to make the hair-puling "intermittent coverage fail" disappear 😉

File: aiosmtpd/tests/test_starttls.py
Comment (pepoluan) [inline aiosmtpd/tests/test_starttls.py:108]: Without this line, `SMTP.session` does not get updated.

**Review by pepoluan — COMMENTED** [review]
Final changes, and a "secret sauce" recipe 😁

File: .github/workflows/test.yml
Comment (pepoluan) [reply .github/workflows/test.yml]: Won't it upload the results of all 6 matrices, though? I don't know what will happen if we do that...

But it's definitely simpler and probably more reliable than making `bash` run a script pulled from the Internet.

I'll implement the change in a new commit.

File: .github/workflows/unit-testing-and-coverage.yml
Comment (pepoluan) [inline .github/workflows/unit-testing-and-coverage.yml:53]: As suggested by @asvetlov in https://github.com/aio-libs/aiosmtpd/pull/211#discussion_r528799855

**Review by pepoluan — COMMENTED** [review]
No additional comment, just ack-ing the new(est) changes.

**Comment by pepoluan** [comment]
If there is no objections @waynew @asvetlov I'll merge this.

**Review by asvetlov — APPROVED** [review]
Please go ahead

**Comment by waynew** [comment]
🚀

**Pull request merged**
